File: server/src/main/java/by/brawl/ws/holder/HeroHolder.java
package by.brawl.ws.holder;

import by.brawl.entity.Hero;
import by.brawl.util.Mappers;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class HeroHolder {
    private String id;
    private Set<SpellHolder> spells = new LinkedHashSet<>();
    private Set<BodypartHolder> bodyparts = new LinkedHashSet<>();
    private Set<EquipmentHolder> equipments = new LinkedHashSet<>();
    private HeroAttributesHolder attributes;

    public HeroHolder(Hero hero) {
        id = hero.getId();
        spells = Mappers.asSet(hero.getSpells(), SpellHolder::new);
        bodyparts = Mappers.asSet(hero.getBodyparts(), BodypartHolder::new);
        equipments = Mappers.asSet(hero.getEquipments(), EquipmentHolder::new);
        attributes = new HeroAttributesHolder(hero.getBodyparts());
    }

    public Set<SpellHolder> getAvailableSpells() {
        return spells.stream()
                .filter(SpellHolder::isAvailable)
                .collect(Collectors.toSet());
    }

    public Set<SpellHolder> getAllSpells() {
        return spells;
    }

    public void heal(Integer value) {
        attributes.getHealth().incrementCurrent(value);
    }

    public void hit(Integer value) {
        attributes.getHealth().decrementCurrent(value);
    }

    public Boolean isAlive() {
        return attributes.getHealth().getCurrent() > 0;
    }

    public String getId() {
        return id;
    }

    public Set<BodypartHolder> getBodyparts() {
        return bodyparts;
    }

    public Set<EquipmentHolder> getEquipments() {
        return equipments;
    }

    public HeroAttributesHolder getAttributes() {
        return attributes;
    }
}

File: server/src/main/java/by/brawl/ws/holder/HeroHolder.kt
package by.brawl.ws.holder

import by.brawl.entity.Hero
import by.brawl.ws.huihui.SpellsPool
import by.brawl.ws.huihui.effects.EffectHolder
import java.util.*

class HeroHolder(hero: Hero,
                 private val squadHolder: SquadHolder) {

    val id: String = hero.id!!
    val name: String = hero.name
    val allSpells = hero.spells.map {
        SpellHolder(SpellsPool.spellsMap[it.id]
                ?: throw IllegalStateException("Spell with id ${it.id} isn't present in spell pool."))
    }.toList()
    val bodyparts = hero.bodyparts.map { BodypartHolder(it) }.toCollection(LinkedHashSet<BodypartHolder>())
    val equipments = hero.equipments.map { EquipmentHolder(it) }.toCollection(LinkedHashSet<EquipmentHolder>())
    val attributes = HeroAttributesHolder(hero.bodyparts)
    val effects = mutableListOf<EffectHolder>()

    fun isAlive() = attributes.health.current > 0

    fun heal(value: Int) {
        attributes.health.incrementCurrent(value)
    }

    fun hit(value: Int) {
        if (value > attributes.barrier.current) {
            val delta = value - attributes.barrier.current
            attributes.barrier.decrementCurrent(delta)
            attributes.health.decrementCurrent(value - delta)
        } else {
            attributes.barrier.decrementCurrent(value)
        }
    }

    fun barrier(value: Int) {
        attributes.barrier.incrementCurrent(value)
    }

    fun position(): Int = squadHolder.heroes.indexOf(this)

    fun move(value: Int) {
        val oldPosition = position()
        var newPosition = oldPosition + value
        if (newPosition < 0) {
            newPosition = 0
        } else if (newPosition > squadHolder.heroes.size - 1) {
            newPosition = squadHolder.heroes.size - 1
        }
        if (oldPosition <= newPosition) {
            Collections.rotate(squadHolder.heroes.subList(oldPosition, newPosition + 1), -1);
        } else {
            Collections.rotate(squadHolder.heroes.subList(oldPosition, newPosition + 1), 1);
        }
    }
}